File: ASP.NET MVC. RedirectToAction propagando el QueryString/[VB]-ASP.NET MVC. RedirectToAction propagando el QueryString/VB/RedirectQueryStringVB/Controllers/HomeController.vb
Imports System.Web.Mvc
Imports System.Web.Routing

Namespace Controllers
    Public Class HomeController
        Inherits Controller

        Function Index() As String
            Dim result As String = "Action: Index<br/>"
            result += "<b>RouteData</b><br/>"
            For Each key As String In RouteData.Values.Keys
                result += String.Format("{0}: {1}<br/>", key, RouteData.Values(key))
            Next
            result += "<b>QueryString</b><br/>"
            For Each key As String In Request.QueryString.AllKeys
                result += String.Format("{0}: {1}<br/>", key, Request.QueryString(key))
            Next
            Return result
        End Function

        Function Redirect() As RedirectToRouteResult
            Dim routeData As RouteValueDictionary = New RouteValueDictionary()
            For Each key As String In Request.QueryString.AllKeys
                routeData.Add(key, Request.QueryString(key))
            Next
            Return RedirectToAction("Index", routeData)
        End Function

        <PropagarQueryString>
        Function RedirectQueryString() As RedirectToRouteResult
            Return RedirectToAction("Index")
        End Function

    End Class
End Namespace

File: ASP.NET MVC. RedirectToAction propagando el QueryString/[C#]-ASP.NET MVC. RedirectToAction propagando el QueryString/C#/RedirectQueryString/Controllers/HomeController.cs
using RedirectQueryString.Code;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using System.Web.Routing;

namespace RedirectQueryString.Controllers
{
    public class HomeController : Controller
    {
        public string Index()
        {
            string result = "Action: Index<br/>";
            result += "<b>RouteData</b><br/>";
            foreach (string key in RouteData.Values.Keys)
                result += string.Format("{0}: {1}<br/>", key, RouteData.Values[key]);
            result += "<b>QueryString</b><br/>";
            foreach (string key in Request.QueryString.AllKeys)
                result += string.Format("{0}: {1}<br/>", key, Request.QueryString[key]);
            return result;
        }

        public RedirectToRouteResult Redirect()
        {
            RouteValueDictionary routeData = new RouteValueDictionary();
            foreach (string key in Request.QueryString.AllKeys)
                routeData.Add(key, Request.QueryString[key]);
            return RedirectToAction("Index", routeData);
        }

        [PropagarQueryString]
        public RedirectToRouteResult RedirectQueryString()
        {
            return RedirectToAction("Index");
        }

    }
}